src/Microsoft.AspNetCore.Blazor/Routing/NavLink.cs

```diff
@@ -100,19 +100,20 @@ private bool ShouldMatch(string currentUriAbsolute)
 
         private void Render(RenderTreeBuilder builder)
         {
-            builder.OpenElement(0, "a");
+            var i = 0;
+            builder.OpenElement(i++, "a");
 
             // Set "active" class dynamically
-            builder.AddAttribute(0, "class", CombineWithSpace(_cssClass, _isActive ? "active" : null));
+            builder.AddAttribute(i++, "class", CombineWithSpace(_cssClass, _isActive ? "active" : null));
 
             // Pass through all other attributes unchanged
-            foreach (var kvp in _allAttributes.Where(kvp => kvp.Key != "class"))
+            foreach (var kvp in _allAttributes.Where(kvp => kvp.Key != "class" && kvp.Key != RenderTreeBuilder.ChildContent))
             {
-                builder.AddAttribute(0, kvp.Key, kvp.Value);
+                builder.AddAttribute(i++, kvp.Key, kvp.Value);
             }
 
             // Pass through any child content unchanged
-            builder.AddContent(1, _childContent);
+            builder.AddContent(i++, _childContent);
```

Comment:
I'm not 100% certain if this is intended to work or not 😆 - I think that's the most important factor for how we proceed.

Reply:
It's not necessary (or shouldn't be) to auto-generate the sequence numbers here. You can keep the sequence number as `0` for the element, and `1` for all the attributes.

Reason:

* The diff algorithm uses the sequence numbers as a basis for a merge-join on sibling nodes. It's slightly more specialised than a normal SQL-style merge join because using the sequence numbers, it can also identify if/else branches and loops. The sequence numbers let us reduce the diff to a linear O(N) operation (versus O(N^3) for general tree diff), and cause the edit script to match very closely the groupings in the original source code (e.g., blocks surrounded by `if`/`else` are atomic). 
* However, the handling for attributes on a given node is different. It's a simpler hash join instead of the loop-aware merge join, because we don't care about the final output order (attributes are independent of order), plus attributes inherently come with string keys (attribute name) and the total number of attributes tends to be small (<10, say). The hash join doesn't compare relative sequence numbers at all. So for attributes, it's fine just to give all of them on a given node the same value. Also it's not harmful to give them distinct values, so we just do whatever's most convenient.

I'm totally aware that the rules here are complex, and we can't expect most users to know about them. People will definitely try to auto-generate sequence numbers at runtime, even though AFAIK there's no scenario where runtime execution order is relevant - it's only source code instruction order that's relevant.

What I'm hoping is that:

1. Most devs will write render methods in the form of Razor files exclusively, so they don't even need to know that sequence numbers exist. I'd hope that even `NavLink` could be written purely in Razor once we get further with #1.
2. For devs building more advanced components who really have to bypass Razor, if they get the sequence number generation wrong, it will still produce a correct result at runtime, even if the diff algorithm produces suboptimal edit scripts.
3. For devs building really advanced components who also need the edit scripts to be optimal, they will have to understand the rules.

What do you think?

Reply:
Thanks for the explanation. If I understand correctly, then there's a bug that I need to investigate then with that you're describing.

Is this expected to work when the element, attributes and content **all** share the sequence number `0`?

Or the numbering then look like:
```
            var i = 0;
            builder.OpenElement(i++, "a");

            // Set "active" class dynamically
            builder.AddAttribute(i, "class", CombineWithSpace(_cssClass, _isActive ? "active" : null));

            // Pass through all other attributes unchanged
            foreach (var kvp in _allAttributes.Where(kvp => kvp.Key != "class" && kvp.Key != RenderTreeBuilder.ChildContent))
            {
                builder.AddAttribute(i, kvp.Key, kvp.Value);
            }

            // Pass through any child content unchanged
            builder.AddContent(++i, _childContent);

            builder.CloseElement();
```

What I was seeing without any code changes to this class is that *all** of the attributes would removed on the second render if the link was not active.

That looks like this.... Initial render
```
<a class="active" href="/Home" childcontent"Microsoft....">
  ...child content...
</a>
```

And then, I navigate to counter, now link is no longer active, and it renders as:
```
<a>
  ...child content...
</a>
```

So once the render encountered the 'remove' attribute edit for `class`, it was also convinced that the other attributes should be removed as well.

We can discuss at the sync up if its not obvious what I'm talking about 😆

Reply:
Ah wait, I was misremembering and mixing up how the original prototype diffing worked compared with how the current one does.

In the current implementation, we *do* use the source sequences for attributes. It's the same function (`AppendDiffEntriesForRange`) that processes all ranges of sibling nodes, where siblings are "nodes with the same parent" and "attributes of the same node".

However I'm still surprised that giving them all the same sequence number causes them to disappear. I'd expect it to produce a valid but not necessarily optimal sequence number. Let's discuss later.